packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -749,6 +754,28 @@ export function resetComponentRoot(vm: VM) {
     vm.velements = EmptyArray;
 }
 
+// Helper function to traverse a tree of VFragment nodes and remove all root children.
+// This is moved into a separate function to minimize the perf/mem impact of the stack traversal
+// on non-vfragment use cases
+function removeFragmentChildren(vnode: VFragment, vm: VM) {
+    const {
+        renderRoot,
+        renderer: { remove },
+    } = vm;
+
+    const nodeStack: VNode[] = [];
+    ArrayPush.call(nodeStack, ...vnode.children);
+
+    let currentNode: VNode | null | undefined;
+    while (!isUndefined((currentNode = ArrayPop.call(nodeStack)))) {
```

Comment:
Use a stack to traverse the potential tree of VFragments to avoid the overhead of functions and objects being created unnecessarily.